Applications/Spire/Source/Ui/ListView.cpp

```diff
@@ -0,0 +1,572 @@
+          get_column_or_row(m_column_or_row_index - 1)->geometry().width());
+        cross_move(false);
+        break;
+      }
+    }
+    update_current(move_previous());
+    break;
+  case Qt::Key_Right:
+    if(m_direction == Qt::Vertical && m_overflow == Overflow::WRAP) {
+      auto column_width =
+        get_column_or_row(m_column_or_row_index)->geometry().width();
+      if(m_tracking_position.x() + column_width <
+          get_column_or_row(get_layout()->count() - 2)->geometry().right()) {
+        m_tracking_position.setX(m_tracking_position.x() + column_width +
+          get_column_or_row(m_column_or_row_index + 1)->geometry().width());
+        cross_move(true);
+        break;
+      }
+    }
+    update_current(move_next());
+    break;
+  default:
+    auto key = event->text();
+    if(!key.isEmpty() && key[0].isLetterOrNumber()) {
+      m_query += key.toLower();
+      m_query_timer.start(QUERY_TIMEOUT_MS);
+      query();
+    }
+    break;
+  }
+  QWidget::keyPressEvent(event);
+}
+
+void ListView::resizeEvent(QResizeEvent* event) {
+  if(event->size() != event->oldSize()) {
+    update_layout();
+  }
+}
+
+scoped_connection ListView::connect_item_current(ListItem* item,
+    const QString& value) {
+  return item->connect_current_signal([=] {
+    if(!m_is_setting_item_focus) {
+      m_current_index = get_index_by_value(value);
+      if(m_current_index == -1) {
+        m_current_model->set_current(boost::none);
+      } else {
+        m_current_model->set_current(value);
+      }
+      update_tracking_position();
+    }
+    m_is_setting_item_focus = false;
+  });
+}
+
+scoped_connection ListView::connect_item_submit(ListItem* item,
+    const QString& value) {
+  return item->connect_submit_signal([=] {
+    m_current_index = get_index_by_value(value);
+    update_tracking_position();
+    if(!m_is_selection_follows_focus) {
+      update_selection(value);
+    }
+    m_submit_signal(value);
+  });
+}
+
+int ListView::get_index_by_value(const QString& value) const {
+  for(auto i = 0; i < m_list_model->get_size(); ++i) {
+    if(m_list_model->get<QString>(i) == value) {
+      return i;
+    }
+  }
+  return -1;
+}
+
+QLayout* Spire::ListView::get_layout() const {
+  return m_body->layout();
+}
+
+QLayoutItem* ListView::get_column_or_row(int index) const {
+  return get_layout()->itemAt(index);
+}
+
+void ListView::select_item(bool is_selected) {
+  for(auto i = 0; i < m_list_model->get_size(); ++i) {
+    if(m_list_model->get<QString>(i) == m_selected) {
+      m_items[i].m_item->set_selected(is_selected);
+    }
+  }
+}
+
+void ListView::cross_move(bool is_next) {
+  m_column_or_row_index = [=] {
+    if(is_next) {
+      return m_column_or_row_index + 2;
+    } else {
+      return m_column_or_row_index - 2;
+    }
+  }();
+  auto target = get_column_or_row(m_column_or_row_index)->layout();
+  auto [v0, min_value] = [=] () -> std::tuple<int, int> {
```

Comment:
Capture by reference.

Reply:
Why is the capture by reference?  The target variable in lambda function is a pointer of QLayout, and m_tracking_position is this->m_tracking_position.

Reply:
Use capture by reference when the lifetime of the closure evaluated by the lambda expression does not outlive the lambda expression's lexical scope.